### PR TITLE
feat: Nonlinear Max Change Reporting to include cell location

### PR DESCRIPTION
**Pull request by tjb-ltk**
Added global in reporting of newton iter  maxpres/temp/compdens changes.  Global cell element index is reported in vtk output to enable finding areas impacting convergence.

### CODE REVIEW

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: got a bit confused here - is there a real need to store all subregions values?

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: might be a bit confusing for a user but up to you - i would use "cell" instead of "global index"

**Review by paveltomin — APPROVED** [review]

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp
Comment (tjb-ltk) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: probably not but there is a subregion loop .. so ??

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp
Comment (tjb-ltk) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: agreed

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: i was thinking something simple like that
```
if (subRegionData.localMaxDeltaPres < maxRegionDeltaPresLoc[1] )
  maxRegionDeltaPresLoc = {subRegionData.localMaxDeltaPres, localToGlobalMap[subRegionData.localMaxDeltaPresLoc};
}
```
?

Comment (tjb-ltk) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: I'll simplify... but thought the max on tuple with lambda function  was pretty cool...

**Comment by paveltomin** [comment]
can we move this to merge queue?

**Comment by paveltomin** [comment]
needs approval from code owner

File: src/coreComponents/common/MpiWrapper.hpp
Comment (CusiniM) [inline src/coreComponents/common/MpiWrapper.hpp]: I think you should add `broadcast` step. I understand that you don't need it for this specific task but since you are creating a generic function I would expect it to also perform the broadcast.

File: src/coreComponents/common/MpiWrapper.hpp
Comment (CusiniM) [inline src/coreComponents/common/MpiWrapper.hpp]: First, I would suggest only passing the buffer to rank 0 with something like:
```suggestion
  MPI_Gather( &tuple, sizeof(tuple_type), MPI_BYTE, myRank == 0 ? l_rcv_buffer.data() : nullptr, sizeof(tuple_type), MPI_BYTE, 0, comm );
```
second,

I think this is actually kind of problematic. I don't think a tuple is always stored as a contiguous block of memory which is what `MPI_Gather` expects since it's passed as MPI_BYTE. I believe that this can lead to weird behaviors if the data is not contiguous. You could copy it into a buffer with contiguous memory and use that to perform the gather but it is somewhat involved. Do we really need this max over tuples? 

@wrtobin and @rrsettgast can probably comment more on this.

**Review by CusiniM — COMMENTED** [review]
I am not sure that the MPI max over the tuple is entirely correct as it is. In any case I don't know that we really need it. I feel we only care about the global max value and its location so we don't need to store the max per subregion IMO.

File: src/coreComponents/common/MpiWrapper.hpp
Comment (CusiniM) [reply src/coreComponents/common/MpiWrapper.hpp]: Also, this really expects a pair not a tuple and I think we should have some checks on the types contained in the tuple. I.e. The first type must be comparable.

Comment (rrsettgast) [reply src/coreComponents/common/MpiWrapper.hpp]: It is not guaranteed by the standard. The tuple should be explicitly packed/unpacked.

File: src/coreComponents/common/MpiWrapper.hpp
Comment (rrsettgast) [reply src/coreComponents/common/MpiWrapper.hpp]: Either broadcast, or change the name of the function.

**Review by rrsettgast — COMMENTED** [review]
I wonder if it would be better to create a class for what you are storing in the `tuple` in this PR? That way it would be guaranteed contiguous, and any operations on the data in the `tuple` could be put into type safe functions

**Comment by tjb-ltk** [comment]
I'm changing from tuple to struct...and will broadcast...

File: src/coreComponents/common/MpiWrapper.hpp
Comment (wrtobin) [reply src/coreComponents/common/MpiWrapper.hpp]: Yeah the types contained within the tuple would all need to be standard layout IIRC for this to be valid. Packing/unpacking is the easier lift than enforcing it through type trait checks.

Comment (tjb-ltk) [reply src/coreComponents/common/MpiWrapper.hpp]: submitted change to use struct for storage and serialization to/from byte string  for mpiallgather

File: src/coreComponents/common/MpiWrapper.hpp
Comment (CusiniM) [inline src/coreComponents/common/MpiWrapper.hpp]:
```suggestion
  template< typename CONTAINER_TYPE > static CONTAINER_TYPE maxValLoc( CONTAINER_TYPE l_valLoc, MPI_Comm comm = MPI_COMM_GEOS );
```
might be better?

File: src/coreComponents/common/MpiWrapper.hpp
Comment (CusiniM) [inline src/coreComponents/common/MpiWrapper.hpp]:
```suggestion
```

File: src/coreComponents/common/MpiWrapper.hpp
Comment (CusiniM) [inline src/coreComponents/common/MpiWrapper.hpp]:
```suggestion
  if constexpr ( std::is_pointer_v< decltype(&valLoc_type::value) > || std::is_pointer_v< decltype(&valLoc_type::location) > ) {
    GEOS_ERROR( "Attempting to use pointer for maxValLoc comparison" );
  }
```

File: src/coreComponents/common/MpiWrapper.hpp
Comment (CusiniM) [inline src/coreComponents/common/MpiWrapper.hpp]: shouldn't this be:
```suggestion
  valLoc_type maxValLoc= *std::max_element( begin( g_valLoc ), end( g_valLoc ), []( auto & lhs, auto & rhs ) -> bool {return lhs.value  <  rhs.value; } );
```
to me it looks like `std::max_element` expects a bool from the comparison: https://en.cppreference.com/w/cpp/algorithm/max_element

File: src/coreComponents/common/MpiWrapper.hpp
Comment (CusiniM) [inline src/coreComponents/common/MpiWrapper.hpp]: can we use an `std:vector` of `unsigned char` ? That one would not require us to manually delete the pointers.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp]:
```suggestion
  template< typename VALUE_TYPE, typename INDEX_TYPE >
```

**Review by CusiniM — APPROVED** [review]
Just a few cosmetic changes.

**Comment by paveltomin** [comment]
still needs review/reviews from @MelReyCG, @corbett5, @rrsettgast, and/or @wrtobin

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp]:
```suggestion
    VALUE_TYPE value;
    INDEX_TYPE location;
```

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp]:
```suggestion
  struct ValueAndLocation
```

File: src/coreComponents/common/MpiWrapper.hpp
Comment (tjb-ltk) [reply src/coreComponents/common/MpiWrapper.hpp]: already done

File: src/coreComponents/common/MpiWrapper.hpp
Comment (tjb-ltk) [reply src/coreComponents/common/MpiWrapper.hpp]: already done

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp
Comment (tjb-ltk) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp]: already done

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp
Comment (tjb-ltk) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp]: submitted

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp
Comment (tjb-ltk) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.hpp]: submitted

**Comment by CusiniM** [comment]
Missing a review. @rrsettgast can you approve it?

**Comment by paveltomin** [comment]
Waiting on code owner review from @MelReyCG, @corbett5, @rrsettgast, and/or @wrtobin.

File: src/coreComponents/common/MpiWrapper.hpp
Comment (corbett5) [inline src/coreComponents/common/MpiWrapper.hpp:1130]: There's certainly a more efficient way to do this, but my guess is this isn't in a performance critical section so it doesn't matter.

**Review by corbett5 — APPROVED** [review]

**Pull request merged**
